File: js/curveFit.js
/* js/curveFit.js — fits a visitor's own ratings against the aggregated Nolan
   Index metric. Mirrors src/personalization/curve_fit.py exactly — change
   both or neither. */

const MIN_POINTS_FOR_DEGREE = { 1: 3, 2: 4, 3: 5 };
const CUBIC_MIN_POINTS = 8;

/* ── Small linear-algebra helpers (matrices are at most 4x4 here) ── */

function transpose(A) {
  return A[0].map((_, j) => A.map((row) => row[j]));
}

function matMul(A, B) {
  return A.map((row) => B[0].map((_, j) => row.reduce((s, v, k) => s + v * B[k][j], 0)));
}

function matVec(A, v) {
  return A.map((row) => row.reduce((s, a, i) => s + a * v[i], 0));
}

/** Gaussian elimination with partial pivoting. */
function solve(A, b) {
  const n = b.length;
  const M = A.map((row, i) => [...row, b[i]]);
  for (let col = 0; col < n; col++) {
    let pivot = col;
    for (let r = col + 1; r < n; r++) {
      if (Math.abs(M[r][col]) > Math.abs(M[pivot][col])) pivot = r;
    }
    [M[col], M[pivot]] = [M[pivot], M[col]];
    for (let r = 0; r < n; r++) {
      if (r === col) continue;
      const factor = M[r][col] / M[col][col];
      for (let c = col; c <= n; c++) M[r][c] -= factor * M[col][c];
    }
  }
  return M.map((row, i) => row[n] / row[i]);
}

/** Least-squares polynomial fit via the normal equations. Highest power first. */
function polyfit(x, y, degree) {
  const n = degree + 1;
  const V = x.map((xi) => Array.from({ length: n }, (_, k) => xi ** (n - 1 - k)));
  const VT = transpose(V);
  return solve(matMul(VT, V), matVec(VT, y));
}

function evaluate(coeffs, x) {
  return coeffs.reduce((acc, c) => acc * x + c, 0);
}

function r2(yTrue, yPred) {
  const mean = yTrue.reduce((s, v) => s + v, 0) / yTrue.length;
  const ssRes = yTrue.reduce((s, v, i) => s + (v - yPred[i]) ** 2, 0);
  const ssTot = yTrue.reduce((s, v) => s + (v - mean) ** 2, 0);
  return ssTot > 1e-9 ? 1 - ssRes / ssTot : 0;
}

/* Leave-one-out R², scored only on held-out points the remaining points
   bracket. A film outside the training x-range (usually Tenet) can only be
   extrapolated, and no polynomial can be fairly validated on extrapolation —
   the catastrophic held-out error would force a flat linear fit on exactly
   the users whose taste is most curved. Extreme films still train every
   other fold. Predictions are clamped, as deployed. */
function looR2(x, y, degree) {
  const preds = [], trues = [];
  for (let i = 0; i < x.length; i++) {
    const xi = x.filter((_, j) => j !== i);
    const yi = y.filter((_, j) => j !== i);
    if (x[i] < Math.min(...xi) || x[i] > Math.max(...xi)) continue;
    preds.push(clampRating(evaluate(polyfit(xi, yi, degree), x[i])));
    trues.push(y[i]);
  }
  return r2(trues, preds);
}

function candidateDegrees(n) {
  let degrees = Object.entries(MIN_POINTS_FOR_DEGREE)
    .filter(([, min]) => n >= min)
    .map(([d]) => Number(d));
  if (n < CUBIC_MIN_POINTS) degrees = degrees.filter((d) => d !== 3);
  return degrees;
}

export function clampRating(raw) {
  return Math.min(5, Math.max(0.5, raw));
}

/** Fits the best-generalizing polynomial (by leave-one-out R²) through (x, y) pairs. */
export function fitBestCurve(x, y) {
  if (x.length < 3) throw new Error("Need at least 3 rated films to fit a curve.");

  let best = null;
  for (const degree of candidateDegrees(x.length)) {
    const r2Loo = looR2(x, y, degree);
    if (!best || r2Loo > best.r2Loo) best = { degree, r2Loo };
  }

  const coeffs = polyfit(x, y, best.degree);
  const r2InSample = r2(y, x.map((xi) => clampRating(evaluate(coeffs, xi))));
  return { degree: best.degree, coeffs, r2InSample, r2Loo: best.r2Loo };
}

export function predictFromFit(fit, x) {
  return clampRating(evaluate(fit.coeffs, x));
}

File: src/personalization/curve_fit.py
"""Fits one person's ratings against the aggregated Nolan Index metric.

Tries a small family of polynomial degrees and picks whichever generalizes
best under leave-one-out cross-validation, rather than whichever fits the
training points most tightly — in-sample R² trivially favors the highest
degree available, which with ~12 points per person would mean nonsense.
"""
import warnings
from dataclasses import dataclass

import numpy as np

# Minimum points needed for a degree-d fit to still have 2+ points left out
# one at a time during LOO-CV.
MIN_POINTS_FOR_DEGREE = {1: 3, 2: 4, 3: 5}
CUBIC_MIN_POINTS = 8   # only offer cubic once there's enough data to trust it


@dataclass
class CurveFit:
    degree: int
    coeffs: np.ndarray      # highest power first (numpy.polyfit convention)
    r2_in_sample: float
    r2_loo: float

    def predict(self, x: float) -> float:
        return float(_clamp(np.polyval(self.coeffs, x)))


def _clamp(prediction):
    return np.clip(prediction, 0.5, 5.0)


def _r2(y_true: np.ndarray, y_pred: np.ndarray) -> float:
    ss_res = float(np.sum((y_true - y_pred) ** 2))
    ss_tot = float(np.sum((y_true - y_true.mean()) ** 2))
    return 1.0 - ss_res / ss_tot if ss_tot > 1e-9 else 0.0


def _loo_r2(x: np.ndarray, y: np.ndarray, degree: int) -> float:
    """Leave-one-out R², scored only on held-out points the remaining points
    bracket. A film outside the training x-range (usually Tenet, far beyond
    everything else on the wackiness axis) can only be extrapolated, and no
    polynomial can be fairly validated on extrapolation — the catastrophic
    held-out error would force a flat linear fit on exactly the users whose
    taste is most curved. Extreme films still train every other fold."""
    preds, trues = [], []
    for i in range(len(y)):
        keep = np.arange(len(y)) != i
        if not (x[keep].min() <= x[i] <= x[keep].max()):
            continue
        # Near-duplicate x values (e.g. films collapsing together at an extreme
        # aggregation weight) make polyfit rank-deficient; those candidates
        # score badly under LOO and lose the selection anyway, so the warning
        # carries no signal.
        with warnings.catch_warnings():
            warnings.simplefilter("ignore", np.exceptions.RankWarning)
            coeffs = np.polyfit(x[keep], y[keep], degree)
        # Score the predictor as deployed — clamped to the rating scale
        preds.append(float(_clamp(np.polyval(coeffs, x[i]))))
        trues.append(y[i])
    return _r2(np.array(trues), np.array(preds))


def _candidate_degrees(n_points: int) -> list[int]:
    degrees = [d for d, minimum in MIN_POINTS_FOR_DEGREE.items() if n_points >= minimum]
    if n_points < CUBIC_MIN_POINTS:
        degrees = [d for d in degrees if d != 3]
    return degrees


def fit_user_curve(metric_values, ratings) -> CurveFit:
    """metric_values, ratings: parallel sequences, one entry per rated film."""
    x = np.array(metric_values, dtype=float)
    y = np.array(ratings, dtype=float)
    if len(x) < 3:
        raise ValueError("Need at least 3 rated films to fit a curve.")

    best_degree, best_r2_loo = None, None
    for degree in _candidate_degrees(len(x)):
        r2_loo = _loo_r2(x, y, degree)
        if best_r2_loo is None or r2_loo > best_r2_loo:
            best_degree, best_r2_loo = degree, r2_loo

    with warnings.catch_warnings():
        warnings.simplefilter("ignore", np.exceptions.RankWarning)
        coeffs = np.polyfit(x, y, best_degree)
    r2_in_sample = _r2(y, _clamp(np.polyval(coeffs, x)))
    return CurveFit(degree=best_degree, coeffs=coeffs,
                    r2_in_sample=r2_in_sample, r2_loo=best_r2_loo)
